contrib/natsio/src/main/java/com/google/cloud/dataflow/contrib/natsio/NatsIO.java

```diff
@@ -0,0 +1,395 @@
+				// Create a NatsSource instance for each split.
+				for(int i = 0; i < numSplits; i++) {
+					results.add(new NatsSource(subject, props));
+				}
+				return results;
+			}
+
+			@Override
+			public Coder<UnboundedSource.CheckpointMark> getCheckpointMarkCoder() {
+				return null;
+			}
+
+			@Override
+			public Coder<KV<String,String>> getDefaultOutputCoder() {
+				return KvCoder.of(NatsIO.DEFAULT_CODER, NatsIO.DEFAULT_CODER);
+			}
+
+			@Override
+			public void validate() {
+				Preconditions.checkNotNull(props.getProperty("servers"), "need to set the 'servers' property");
+				Preconditions.checkNotNull(subject, "need to set the 'subject'");
+			}
+
+			public static class CommitMark implements UnboundedSource.CheckpointMark, Serializable {
+				private static final long serialVersionUID = 2L;
+
+				public void finalizeCheckpoint() throws IOException {}		
+			}
+				
+			private class NatsReader extends UnboundedReader<KV<String,String>> {
+				private final String subject;
+				private Integer sid;
+				private final ConcurrentLinkedQueue<KV<String,String>> queue;
+				private KV<String,String> current;
+				
+				public NatsReader(String subject) {
+					if (subject == null) {
+						throw new NullPointerException("subject");
+					}
+					this.subject = subject;
+					queue = new ConcurrentLinkedQueue<KV<String,String>>();
+				}
+				
+				@Override
+				public boolean advance() throws IOException {
+					// Polling a queue to check whether an update is available in the queue.
+					current = queue.poll();
+					if (current == null) {
+						return false;
+					}
+					
+					return true;
+				}
+
+				@Override
+				public UnboundedSource.CheckpointMark getCheckpointMark() {
+					return DEFAULT_MARK;
+				}
+
+				@Override
+				public byte[] getCurrentRecordId() throws NoSuchElementException {
+					return DEFAULT_ID;
+				}
+
+				@Override
+				public NatsSource getCurrentSource() {
+					return NatsSource.this;
+				}
+
+				@Override
+				public Instant getWatermark() {
+					return Instant.now();
+				}
+
+				@Override
+				public boolean start() throws IOException {
+					if (conn == null) {
+						try {
+							conn = Connection.connect(props);
+						} catch (InterruptedException e) {
+							throw new ClosedByInterruptException();
+						}
+					}
+					
+					// Subscribe to a subject and attach an event handler to store updates in a queue.
+					sid = conn.subscribe(subject, props, new MsgHandler() {
+						@Override
+						public void execute(String msg, String reply, String subject) {
+							queue.add(KV.of(subject, msg));
+						}
+					});
+					return advance();
+				}
+
+				@Override
+				public void close() throws IOException {
+					conn.unsubscribe(sid);
+				}
+
+				@Override
+				public KV<String,String> getCurrent() throws NoSuchElementException {
+					return current;
+				}
+
+				@Override
+				public Instant getCurrentTimestamp() throws NoSuchElementException {
+					return Instant.now();
```

Comment:
This method does not meet its spec. Specifically, if the user calls calls this method twice without an intervening call to `advance()`, the source will return the same element (in `getCurrent`) -- and that element should have the same timestamp.